contributors/design-proposals/live-and-inplace-vertical-scaling.md

```diff
@@ -0,0 +1,325 @@
+
+# Objectives
+
+1. Enable live and in-place resource resizing on a pod.
+2. Add support for live and in-place resource resizing in `StatefulSet` controller.
+
+# API and Usage
+
+To express the policy for resizing in a pod spec, we introduce resource attribute `resizePolicy` with the following choices for value:
+* RestartOnly. (the current behavior, default)
+* LiveResizeable.
+
+This attribute will be available per resource (such as cpu, memory) and so is adequate to indicate whether the workload can handle and prefer a change in each resource’s allocation for it without restarting.
+With potentially multiple containers and multiple resizeable resources for each in a Pod, the response to an update of the pod spec will be determined by the a precedence order among the attribute values with RestartOnly dominating LiveResizeable, i.e., if two resources have been resized in the update to the spec and one of them has a policy of RestartOnly then the pod would be restarted to realize both updates.
+
+We can optionally introduce an action annotation `resizeAction` with following choices for value:
+* Restart. (default)
+* LiveResize.
+* LiveResizePreferred.
+
+If used, this would be included as part of the patch or appropriate update command providing the spec update for the resize.
+It would indicate the preference of user at the time of resize.
+Specifically, Restart for `resizeAction` would indicate the pod be restarted for the corresponding resizing of resource(s), LiveResize would indicate the pod not be restarted the resize be realized live, and LiveResizePreferred would indicate that the resize be realized preferrably live but if that fails for any reason to accomplish it with a restart.
+
+An example of the usage of resizePolicy attribute in a pod spec:
+
+```yaml
+resources:
+    requests:
+        cpu: 100m
+        memory: 1Gi
+    limits:
+        cpu: 1000m
+        memory: 1Gi
+    resizePolicy:
+        cpu: LiveResizeable
+        memory: RestartOnly
+```
+
+For the above example, if there is a change to cpu request or limit it can be vertically scaled only if the memory request and limit remained the same, otherwise the RestartOnly policy for memory would override the policy for CPU, and the Pod (container, if container-alone restart is allowed) would need to be restarted.
+
+An example of change in CPU resource `request` size and usage of `resizeAction` annotation with the updated spec.
+
+// before the change
+```yaml
+resources:
+    requests:
+        cpu: **100m**
+        memory: 1Gi
+    limits:
+        cpu: 1000m
+        memory: 1Gi
+    resizePolicy:
+        cpu: LiveResizeable
+        memory: RestartOnly
+```
+
+// after the change
+```yaml
+annotation:
+resizeAction: LiveResizePreferred
+...
+resources:
+    requests:
+        cpu: **400m**
+        memory: 1Gi
+    limits:
+        cpu: 1000m
+        memory: 1Gi
+    resizePolicy:
+        cpu: LiveResizeable
+        memory: RestartOnly
+```
```

Comment:
I am a little bit confused about the `resizeAction` annotation, under which case should I respect this value? and how to coordinate `resizeAction` with `resizePolicy`?

Reply:
The resizePolicy is a kind of the characteristics of resource that describes whether the resource can be resized live or not. The resizeAction is a way that a user/client want to take when resizing the resource. So, for example, even though CPU resource is specified LiveResizeable, but for some reason, if a user specifies the resizeAction for CPU to Restart,  resizing of CPU is done by restart. Of course, for a resource whose resizePolicy is RestartOnly, its resizeAction cannot be LiveResize/LiveResizePreferred.

Reply:
I see your point, thanks for kindly explanation.